File: app/src/main/java/com/kent/university/privelt/ui/settings/SettingsActivity.java
package com.kent.university.privelt.ui.settings;

import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.Scope;
import com.google.api.services.drive.DriveScopes;
import com.kent.university.privelt.R;
import com.kent.university.privelt.base.GoogleDriveActivity;
import com.kent.university.privelt.base.GoogleDriveListener;
import com.kent.university.privelt.model.Settings;
import com.kent.university.privelt.ui.master_password.MasterPasswordActivity;

import androidx.annotation.Nullable;
import butterknife.BindView;

public class SettingsActivity extends GoogleDriveActivity {

    public static final String ARG_CHANGE_PASSWORD = "ARG_CHANGE_PASSWORD";

    @BindView(R.id.change_password)
    Button button;

    @BindView(R.id.logout)
    Button logout;

    @BindView(R.id.drive)
    Switch drive;

    @BindView(R.id.fileId)
    EditText fileId;

    @BindView(R.id.googleId)
    TextView googleId;

    @BindView(R.id.disconnect)
    ImageView disconnect;

    SettingsViewModel settingsViewModel;

    private Settings settings;

    @Override
    protected void configureDesign(@Nullable Bundle savedInstanceState) {

        button.setOnClickListener(view -> {
            Intent intent = new Intent(this, MasterPasswordActivity.class);
            intent.putExtra(ARG_CHANGE_PASSWORD, true);
            startActivity(intent);
        });

        logout.setOnClickListener(view -> new AlertDialog.Builder(this)
                .setTitle(R.string.log_out)
                .setMessage(R.string.log_out_confirmation)
                .setPositiveButton(R.string.yes, (dialog, which) -> {
                    startActivity(new Intent(this, MasterPasswordActivity.class));
                    this.finish();
                })
                .setNegativeButton(R.string.no, null)
                .show());

        listener = new GoogleDriveListener() {
            @Override
            public void onDownloadSuccess() {

            }

            @Override
            public void onDownloadFailure() {

            }

            @Override
            public void onConnectionSuccess() {
                GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(SettingsActivity.this);

                googleId.setText(getResources().getString(R.string.logged_with, account.getDisplayName()));
                disconnect.setVisibility(View.VISIBLE);
                googleId.setVisibility(View.VISIBLE);
            }
        };

        drive.setOnCheckedChangeListener((compoundButton, b) -> {
            settings.setGoogleDriveAutoSave(b);
            if (b)
                googleDriveConnection();
            settingsViewModel.updateSettings(settings);
        });

        fileId.setOnClickListener(v -> {
            ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText("Copied text", fileId.getText().toString());
            clipboard.setPrimaryClip(clip);
            Toast.makeText(SettingsActivity.this, "Text copied to clipboard", Toast.LENGTH_LONG).show();
        });

        disconnect.setOnClickListener(view -> {

            GoogleSignInOptions signInOptions =
                    new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                            .requestEmail()
                            .requestScopes(new Scope(DriveScopes.DRIVE_FILE))
                            .build();
            GoogleSignInClient googleSignInClient = GoogleSignIn.getClient(this, signInOptions);
            googleSignInClient.signOut();

            settings.setGoogleDriveAutoSave(false);
            settingsViewModel.updateSettings(settings);
        });
        configureViewModel();
        getSettings();
    }

    private void getSettings() {
        settingsViewModel.getSettings().observe(this, this::updateSettings);
    }

    private void updateSettings(Settings settings) {
        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);

        this.settings = settings;
        if (settings == null)
            this.settings = new Settings(false, "");
        drive.setChecked(this.settings.isGoogleDriveAutoSave());
        fileId.setText(this.settings.getGoogleDriveFileID());

        if (account != null) {
            googleId.setText(getResources().getString(R.string.logged_with, account.getDisplayName()));
            disconnect.setVisibility(View.VISIBLE);
            googleId.setVisibility(View.VISIBLE);
        } else {
            this.settings.setGoogleDriveAutoSave(false);
            disconnect.setVisibility(View.GONE);
            googleId.setVisibility(View.GONE);
        }
    }

    @Override
    protected int getActivityLayout() {
        return R.layout.activity_settings;
    }

    @Override
    protected void configureViewModel() {
        settingsViewModel = getViewModel(SettingsViewModel.class);
        settingsViewModel.init();
    }
}

File: app/src/main/java/com/kent/university/privelt/ui/settings/SettingsActivity.kt
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.kent.university.privelt.ui.settings

import android.app.Activity
import android.app.AlertDialog
import android.content.*
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.CompoundButton
import android.widget.Toast
import androidx.lifecycle.Observer
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.Scope
import com.google.android.material.snackbar.Snackbar
import com.google.api.services.drive.DriveScopes
import com.kent.university.privelt.R
import com.kent.university.privelt.base.PDAActivity
import com.kent.university.privelt.base.PDAListener
import com.kent.university.privelt.model.Settings
import com.kent.university.privelt.ui.master_password.MasterPasswordActivity
import kotlinx.android.synthetic.main.activity_settings.*

class SettingsActivity : PDAActivity() {

    override val activityLayout: Int
        get() = R.layout.activity_settings
    private var settingsViewModel: SettingsViewModel? = null
    private var settings: Settings? = null

    //TODO Here we need a recyclerview with each service and a ServiceManager
    override fun configureDesign(savedInstanceState: Bundle?) {
        change_password!!.setOnClickListener {
            val intent = Intent(this, MasterPasswordActivity::class.java)
            intent.putExtra(ARG_CHANGE_PASSWORD, true)
            startActivity(intent)
        }
        logout!!.setOnClickListener {
            AlertDialog.Builder(this)
                    .setTitle(R.string.log_out)
                    .setMessage(R.string.log_out_confirmation)
                    .setPositiveButton(R.string.yes) { _: DialogInterface?, _: Int ->
                        startActivity(Intent(this, MasterPasswordActivity::class.java))
                        setResult(Activity.RESULT_OK)
                        finish()
                    }
                    .setNegativeButton(R.string.no, null)
                    .show()
        }
        listener = object : PDAListener {
            override fun onDownloadSuccess() {}
            override fun onDownloadFailure() {}
            override fun onConnectionSuccess() {
                val account = GoogleSignIn.getLastSignedInAccount(this@SettingsActivity)
                googleId!!.text = resources.getString(R.string.logged_with, account!!.displayName)
                disconnect!!.visibility = View.VISIBLE
                googleId!!.visibility = View.VISIBLE
            }

            override fun onHatUploadSuccess(fileId: String) {
                if (settings?.hatFileID != null && settings!!.hatFileID.isNotEmpty())
                    deleteFileOnHAT(settings!!.hatFileID)
                settings!!.hatFileID = fileId
                fileIdEditTextHat.setText(fileId)
                settingsViewModel!!.updateSettings(settings)
            }

            override fun onHatUploadFailure(error: String) {
            }
        }
        drive!!.setOnCheckedChangeListener { _: CompoundButton?, b: Boolean ->
            settings!!.isGoogleDriveAutoSave = b
            if (b) googleDriveConnection()
            settingsViewModel!!.updateSettings(settings)
        }
        fileIdEditTextDrive?.setOnClickListener {
            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText("Copied text", fileIdEditTextDrive?.text.toString())
            clipboard.setPrimaryClip(clip)
            Toast.makeText(this@SettingsActivity, "Text copied to clipboard", Toast.LENGTH_LONG).show()
        }
        fileIdEditTextHat?.setOnClickListener {
            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText("Copied text", fileIdEditTextHat?.text.toString())
            clipboard.setPrimaryClip(clip)
            Toast.makeText(this@SettingsActivity, "Text copied to clipboard", Toast.LENGTH_LONG).show()
        }
        disconnect!!.setOnClickListener {
            val signInOptions = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                    .requestEmail()
                    .requestScopes(Scope(DriveScopes.DRIVE_FILE))
                    .build()
            val googleSignInClient = GoogleSignIn.getClient(this, signInOptions)
            googleSignInClient.signOut()
            settings!!.isGoogleDriveAutoSave = false
            settingsViewModel!!.updateSettings(settings)
        }
        configureViewModel()
        getSettings()
        configureHat()
    }

    private fun getSettings() {
        settingsViewModel!!.settings?.observe(this, Observer { settings: Settings? -> updateSettings(settings) })
    }

    private fun updateSettings(settings: Settings?) {
        val account = GoogleSignIn.getLastSignedInAccount(this)
        this.settings = settings
        if (settings == null) this.settings = Settings(false, "", "")
        drive!!.isChecked = this.settings!!.isGoogleDriveAutoSave
        fileIdEditTextDrive?.setText(this.settings!!.googleDriveFileID)
        if (account != null) {
            googleId!!.text = resources.getString(R.string.logged_with, account.displayName)
            disconnect!!.visibility = View.VISIBLE
            googleId!!.visibility = View.VISIBLE
        } else {
            this.settings!!.isGoogleDriveAutoSave = false
            disconnect!!.visibility = View.GONE
            googleId!!.visibility = View.GONE
        }
        if (settings?.hatFileID != null && settings.hatFileID.isNotEmpty())
            fileIdEditTextHat.setText(settings.hatFileID)
    }

    override fun configureViewModel() {
        settingsViewModel = getViewModel(SettingsViewModel::class.java)
        settingsViewModel?.init()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.help_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.help) {
            Snackbar.make(logout!!, R.string.save_data, Snackbar.LENGTH_LONG).show()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        const val ARG_CHANGE_PASSWORD = "ARG_CHANGE_PASSWORD"
    }

    private fun configureHat() {
        initHatLoginButton()
    }

    private fun initHatLoginButton() {
        hatSwitch.setOnClickListener {
            loginHAT(hatEmail.text.toString())
        }
    }

    override fun hatLogged() {
        uploadDatabaseWithHAT()
    }
}